### PR TITLE
Smartbox implementation

### PR DESCRIPTION
**Pull request by dparikh**
Initial implementation.

### CODE REVIEW

File: examples/sample-pub/views/article.html [hunk 1]
```diff
@@ -110,6 +110,7 @@ <h1 class="heading"><% article.title %></h1>
           <p class="byline">By <% article.author %> on <% article.date %></p>
           <div class="hr"></div>
 
+          <div id="swgId" style="display:none;height:10px;min-width: 20px;background-color: #ff1"></div>
```
Comment (dparikh) [inline examples/sample-pub/views/article.html]: Publisher would require to specify the container `DIV`, where `smartbutton` would be injected by the API call. 

Need to finalize, what `css` `style` attributes are allowed/accepted by SwG.

File: src/api/subscriptions.js [hunk 2]
```diff
@@ -375,6 +386,8 @@ export let SaveSubscriptionRequestCallback;
  *
  * @typedef {{
  *   theme: string,
+ *   lang: (string|undefined),
+ *   isClosable: (boolean|undefined),
```
Comment (dparikh) [inline src/api/subscriptions.js]: `isClosable` will be passed to the `showOffers` API call when `dialog` is opened.

File: src/components/dialog-manager.js [hunk 3]
```diff
@@ -72,17 +72,19 @@ export class DialogManager {
   /**
    * @param {!./view.View} view
    * @param {boolean=} hidden
+   * @param {?Element=} elem
    * @return {!Promise}
    */
-  openView(view, hidden = false) {
+  openView(view, hidden = false, elem) {
```
Comment (dparikh) [inline src/components/dialog-manager.js]: For Testing purpose only.

Dialog will not be used for `smartbutton`

File: src/components/dialog.js [hunk 4]
```diff
@@ -273,14 +273,20 @@ export class Dialog {
   /**
    * Opens the given view and removes existing view from the DOM if any.
    * @param {!./view.View} view
+   * @param {?Element=} elem
    * @return {!Promise}
    */
-  openView(view) {
+  openView(view, elem) {
```
Comment (dparikh) [inline src/components/dialog.js]: For testing purpose only.

File: src/runtime/runtime.js [hunk 5]
```diff
@@ -816,6 +824,31 @@ export class ConfiguredRuntime {
...
+    const opts = /** @type {!ButtonOptions} */
+        (Object.assign(options || {}, {theme: Theme.LIGHT}));
+    if (!isExperimentOn(this.win_, ExperimentFlags.SMARTBOX)) {
+      throw new Error('Not yet launched!');
+    }
+    if (!container) {
+      throw new Error('No element found to render Smart button!');
+    }
+    return this.documentParsed_.then(() => {
+      return this.getEntitlements().then(entitlements => {
```
Comment (dparikh) [inline src/runtime/runtime.js]: `getEntitlements` API call is required to find if there are `subscriptions` and if any, do not display `smartbutton.
Also, `isReadyToPay` flag is passed to the `SmartSubscriptionButtonFlow` to decide what `message` is rendered underneath the button.

File: src/api/subscriptions.js [hunk 6]
```diff
@@ -236,6 +236,16 @@ export class Subscriptions {
...
 
+  /**
+   * Creates smartButton element and the provided callback.
+   * The default theme is "light".
+   *
+   * @param {!Element} container
+   * @param {!ButtonOptions=} opt_options
+   * @return {!Element}
+   */
+  createSubscriptionButton(container, opt_options) {}
```
Comment (dparikh) [inline src/api/subscriptions.js]: rename to `renderSmartButton(container, options, callback)`

File: src/api/subscriptions.js [hunk 7]
```diff
@@ -236,6 +236,17 @@ export class Subscriptions {
...
+  /**
+   * Creates smartButton element and the provided callback.
+   * The default theme is "light".
+   *
+   * @param {!Element} container
+   * @param {!ButtonOptions=} opt_options
+   * @param {function()=} opt_callback
+   * @return {!Promise}
+   */
+  renderSmartButton(container, opt_options, opt_callback) {}
```
Comment (dvoytenko) [inline src/api/subscriptions.js]: Let's also discuss the naming of `attachSmartButton`. The example above is pretty similar from many points of view. We might have a more obvious API if we move in that path.

File: src/api/subscriptions.js [hunk 8]
```diff
@@ -257,6 +268,7 @@ export const SubscriptionFlows = {
   LINK_ACCOUNT: 'linkAccount',
   SHOW_LOGIN_PROMPT: 'showLoginPrompt',
   SHOW_LOGIN_NOTIFICATION: 'showLoginNotification',
+  SMART_SUBSCRIPTION_BUTTON: 'showSmartSubscriptionButton',
```
Comment (dvoytenko) [inline src/api/subscriptions.js]: I'm not convinced we need this.

File: src/components/view.js [hunk 9]
```diff
@@ -41,6 +41,13 @@ export class View {
    */
   init(unusedDialog) {}
 
+
```
Comment (dvoytenko) [inline src/components/view.js]: I'd hope no changes to the `View` interface would be needed.

File: examples/sample-pub/views/article.html [hunk 10]
```diff
@@ -110,6 +110,7 @@ <h1 class="heading"><% article.title %></h1>
           <p class="byline">By <% article.author %> on <% article.date %></p>
           <div class="hr"></div>
 
+          <div id="swgId" style="min-height:126px;min-width: 280px;"></div>
```
Comment (dparikh) [inline examples/sample-pub/views/article.html]: To revert this change before merge.

File: src/runtime/button-api.js [hunk 11]
```diff
@@ -17,6 +17,14 @@
 import {createElement} from '../utils/dom';
 import {msg} from '../utils/i18n';
 
+/**
+ * @enum {string}
+ */
+export const Theme = {
+  'LIGHT': 'light',
```
Comment (dvoytenko) [inline src/runtime/button-api.js]: Enum names should not have quotes.

File: src/runtime/runtime.js [hunk 12]
```diff
@@ -816,6 +824,23 @@ export class ConfiguredRuntime {
     this.buttonApi_.attach(button, optionsOrCallback, opt_callback);
   }
 
+  /** @override */
+  renderSmartButton(container, options, callback) {
+    if (!options) {
+      Object.assign({}, {theme: Theme.LIGHT, lang: 'en'});
```
Comment (dvoytenko) [inline src/runtime/runtime.js]: Hmm. I don't think this is doing anything. Plus the `if (!options)` is also wrong I believe. Per method's signature, the you should be expecting one one of these shapes: 

```
renderSmartButton(container, callback)
```
or
```
renderSmartButton(container, options, callback)
```

Thus, the second argument is not guaranteed to be an options map. And it's never not specified.

File: src/api/subscriptions.js [hunk 13]
```diff
@@ -236,6 +236,17 @@ export class Subscriptions {
...
   attachButton(button, optionsOrCallback, opt_callback) {}
 
+  /**
+   * Creates smartButton element and the provided callback.
+   * The default theme is "light".
+   *
+   * @param {!Element} container
+   * @param {!ButtonOptions=} opt_options
+   * @param {function()=} opt_callback
+   * @return {!Promise}
```
Comment (dvoytenko) [inline src/api/subscriptions.js]: What does this "promise" mean at this time?

File: src/runtime/runtime.js [hunk 14]
```diff
@@ -816,6 +824,23 @@ export class ConfiguredRuntime {
...
+    }
+    if (!isExperimentOn(this.win_, ExperimentFlags.SMARTBOX)) {
+      throw new Error('Not yet launched!');
+    }
+    if (!container) {
+      throw new Error('No element found to render Smart button!');
+    }
+    return this.documentParsed_.then(() => {
+      return new SmartSubscriptionButtonApi(this, container, options, callback)
```
Comment (dvoytenko) [inline src/runtime/runtime.js]: Is this a good idea to return in an API?

File: src/runtime/smart-button-api.js [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+import {ActivityIframeView} from '../ui/activity-iframe-view';
+import {Theme} from './button-api';
+//import {SubscriptionFlows} from '../api/subscriptions';
```
Comment (dvoytenko) [inline src/runtime/smart-button-api.js]: remove

File: src/runtime/smart-button-api.js [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+    this.container_ = container;
+
+    /** @private const {?function()=} */
+    this.callback_ = callback;
+
+    /** @private @const {string} */
+    this.theme_ = options && options.theme || Theme.LIGHT;
+
+    /** @private {boolean} */
+    this.isClosable_ = options && options.isClosable || false;
```
Comment (dvoytenko) [inline src/runtime/smart-button-api.js]: Remove?

File: src/runtime/smart-button-api.js [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+    if (this.isClosable_ == undefined) {
+      this.isClosable_ = false;  // Default is to hide Close button.
+    }
+
+    /** @private @const {!ActivityIframeView} */
+    this.activityIframeView_ = new ActivityIframeView(
```
Comment (dvoytenko) [inline src/runtime/smart-button-api.js]: Can we not have it to be a view right away?

File: src/runtime/smart-button-api.js [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+          'productId': deps.pageConfig().getProductId(),
+          'publicationId': deps.pageConfig().getPublicationId(),
+          'theme': this.theme_,
+          'lang': options && options.lang || 'en',
+        }),
+        /* shouldFadeBody */ false);
+  }
+
+  /**
+   * Starts the smart button subscription button flow.
```
Comment (dvoytenko) [inline src/runtime/smart-button-api.js]: Nit: not a flow.

File: src/runtime/smart-button-api.js [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+  }
+
+  /**
+   * @private
+   */
+  buildContent_() {
+    this.container_.appendChild(this.activityIframeView_.getElement());
+
+    setImportantStyles(this.container_, {
+      'height': '126px',
```
Comment (dvoytenko) [inline src/runtime/smart-button-api.js]: This seems odd to set it on the container. Why not just set it on the element itself? We are doing it early enough so that the developer can reflect the changes early in the development cycle.

File: src/runtime/smart-button-api.js [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+    });
+
+    this.activityIframeView_.initContainer().then(() => {
+      setImportantStyles(this.activityIframeView_.getElement(), {
+        'opacity': 1,
+        'height': '100%',
```
Comment (dvoytenko) [inline src/runtime/smart-button-api.js]: Ditto: this is rather dengerous.

File: src/runtime/button-api.js (hunk 11 above)
Comment (dparikh) [reply src/runtime/button-api.js]: Done

File: src/api/subscriptions.js (hunk 8 above)
Comment (dparikh) [reply src/api/subscriptions.js]: This was left over from previous implementation. Was not being used anywhere, removed.

File: src/runtime/smart-button-api.js (hunk 16 above)
Comment (dparikh) [reply src/runtime/smart-button-api.js]: Removed

File: src/runtime/smart-button-api.js (hunk 18 above)
Comment (dparikh) [reply src/runtime/smart-button-api.js]: Done

File: src/runtime/smart-button-api.js (hunk 15 above)
Comment (dparikh) [reply src/runtime/smart-button-api.js]: Removed

File: src/api/subscriptions.js (hunk 13 above)
Comment (dparikh) [reply src/api/subscriptions.js]: Not required, removed.

File: src/runtime/runtime.js (hunk 14 above)
Comment (dparikh) [reply src/runtime/runtime.js]: Since now it is an API, return is not required.  Removed.

File: src/runtime/smart-button-api.js (hunk 19 above)
Comment (dparikh) [reply src/runtime/smart-button-api.js]: Removed from here. 
We might still want to override any styles, publishers might have set, to ensure that the height is at least (`min-height: 126px;` )

File: src/components/view.js (hunk 9 above)
Comment (dparikh) [reply src/components/view.js]: `view` interface has signature, `init(unusedDialog)` and for `smartButton` we do not have `dialog` param to pass.

File: src/runtime/smart-button-api.js (hunk 20 above)
Comment (dparikh) [reply src/runtime/smart-button-api.js]: Removed.

File: src/components/view.js (hunk 9 above)
Comment (dvoytenko) [reply src/components/view.js]: `View` is still defined as a "view" inside a dialog. So using it w/o a dialog does not seem to be like a right use.

File: src/runtime/smart-button-api.js (hunk 19 above)
Comment (dvoytenko) [reply src/runtime/smart-button-api.js]: Setting it on the element itself is totally ok, i believe. Our main question - how early are we doing this. Here . you do it pretty early, so there's no reason to just set it on your button element.

Comment (dvoytenko) [reply src/runtime/smart-button-api.js]: Still a lot of conflation between a container and a button concepts. I think we should review the `attachSmartButton` path - it could clarify these things. Did we ever try to place our `<iframe>` inside a `<button>`?

File: src/runtime/runtime.js (hunk 12 above)
Comment (dparikh) [reply src/runtime/runtime.js]: Changed the supported signature for `renderSmartButton`

File: src/runtime/smart-button-api.js (hunk 17 above)
Comment (dparikh) [reply src/runtime/smart-button-api.js]: Moved out from constructor.

File: src/components/view.js (hunk 9 above)
Comment (dparikh) [reply src/components/view.js]: OK. Reverting this change.
Making direct call to `activityPort.openIframe` now.

File: src/runtime/smart-button-api.js (hunk 19 above)
Comment (dparikh) [reply src/runtime/smart-button-api.js]: Will discuss offline...

File: src/api/subscriptions.js (hunk 7 above)
Comment (dparikh) [reply src/api/subscriptions.js]: Done as discussed offline

File: examples/sample-pub/views/article.html [hunk 21]
```diff
@@ -110,6 +110,7 @@ <h1 class="heading"><% article.title %></h1>
           <p class="byline">By <% article.author %> on <% article.date %></p>
           <div class="hr"></div>
 
+          <button class="swg-smart-button"></button>
```
Comment (dparikh) [inline examples/sample-pub/views/article.html]: Remove before merge.

File: assets/swg-button.css [hunk 22]
```diff
@@ -257,3 +257,17 @@
 .swg-button-dark:lang(zh-tw):after {
   background-image: url("i18n/b-zh-tw-dk.svg");
 }
+
+.swg-smart-button {
```
Comment (dvoytenko) [inline assets/swg-button.css]: Please add a comment here that all "button" styles will be defined inside the iframe.

File: assets/swg-button.css [hunk 23]
```diff
@@ -257,3 +257,17 @@
...
+  border: none;
+  padding: 0;
+  background: transparent;
+  min-height: 126px;
+  position: relative;
+  width: 300px;
```
Comment (dvoytenko) [inline assets/swg-button.css]: Let's also define a min-width.

File: assets/swg-button.css [hunk 24]
```diff
@@ -257,3 +257,17 @@
 .swg-button-dark:lang(zh-tw):after {
   background-image: url("i18n/b-zh-tw-dk.svg");
 }
+
+.swg-smart-button {
+  border: none;
+  padding: 0;
+  background: transparent;
+  min-height: 126px;
+  position: relative;
```
Comment (dvoytenko) [inline assets/swg-button.css]: Should we also do `overflow: hidden`?

File: assets/swg-button.css [hunk 25]
```diff
@@ -257,3 +257,17 @@
 .swg-button-dark:lang(zh-tw):after {
   background-image: url("i18n/b-zh-tw-dk.svg");
 }
+
```
Comment (dvoytenko) [inline assets/swg-button.css]: Please move it up before all the language re-definitions.

File: examples/sample-pub/publisher.js [hunk 26]
```diff
@@ -104,6 +104,24 @@ function startFlowAuto() {
     });
     return;
   }
+
+  if (flow == 'smartbutton') {
+    whenReady(function(subsciptions) {
+      var subs = subsciptions;
+      whenDemoReady(function() {
+        var smartButtonContainerDiv =
```
Comment (dvoytenko) [inline examples/sample-pub/publisher.js]: It's no longer a container, so let's also remove "container" from the name. And also "div".

File: src/api/subscriptions.js [hunk 27]
```diff
@@ -236,6 +236,16 @@ export class Subscriptions {
    */
   attachButton(button, optionsOrCallback, opt_callback) {}
 
+  /**
+   * Creates smartButton element and the provided callback.
```
Comment (dvoytenko) [inline src/api/subscriptions.js]: s/Creates/Attaches

File: src/runtime/button-api.js [hunk 28]
```diff
@@ -98,20 +100,17 @@ export class ButtonApi {
...
-    const options =
-        typeof optionsOrCallback != 'function' ?
-        optionsOrCallback : null;
-    const callback = /** @type {function()} */ (
-        (typeof optionsOrCallback == 'function' ? optionsOrCallback : null) ||
-            opt_callback);
+    const options = this.getOptions_(optionsOrCallback);
+    const callback = /** @type {function(Event):boolean} */
+        (this.getCallback_(optionsOrCallback, opt_callback));
     let theme = options && options['theme'];
```
Comment (dvoytenko) [inline src/runtime/button-api.js]: So it looks like now options are never null/undefined here?

File: src/runtime/button-api.js [hunk 29]
```diff
@@ -98,20 +100,17 @@ export class ButtonApi {
...
    */
   attach(button, optionsOrCallback, opt_callback) {
-    const options =
-        typeof optionsOrCallback != 'function' ?
-        optionsOrCallback : null;
-    const callback = /** @type {function()} */ (
-        (typeof optionsOrCallback == 'function' ? optionsOrCallback : null) ||
-            opt_callback);
+    const options = this.getOptions_(optionsOrCallback);
+    const callback = /** @type {function(Event):boolean} */
```
Comment (dvoytenko) [inline src/runtime/button-api.js]: Why does it return boolean? Why is typecast needed here if we have a function to return this function to us?

File: src/runtime/button-api.js [hunk 30]
```diff
@@ -122,4 +121,51 @@ export class ButtonApi {
...
+
+  /**
+   *
+   * @param {?../api/subscriptions.ButtonOptions|function():null} optionsOrCallback
+   * @param {function():null=} opt_callback
+   * @return {function():null|function(Event):boolean}
+   * @private
+   */
+  getCallback_(optionsOrCallback, opt_callback) {
+    const callback = /** @type {function():null|function(Event):boolean} */ (
```
Comment (dvoytenko) [inline src/runtime/button-api.js]: Why the variance with return type? Also return type of `null` is not a clear concept.

File: src/runtime/button-api.js [hunk 31]
```diff
@@ -122,4 +121,51 @@ export class ButtonApi {
...
+   * @param {!Element} button
+   * @param {!../api/subscriptions.ButtonOptions|function()} optionsOrCallback
+   * @param {function():null=} opt_callback
+   * @return {!Element}
+   */
+  attachSmartButton(deps, button, optionsOrCallback, opt_callback) {
+    const options = this.getOptions_(optionsOrCallback);
+    const callback = /** @type {function():null} */
+        (this.getCallback_(optionsOrCallback, opt_callback));
+    let theme = options && options['theme'];
```
Comment (dvoytenko) [inline src/runtime/button-api.js]: The `theme` never used?

File: src/runtime/button-api.js [hunk 32]
```diff
@@ -98,20 +100,17 @@ export class ButtonApi {
...
-    const callback = /** @type {function()} */ (
-        (typeof optionsOrCallback == 'function' ? optionsOrCallback : null) ||
-            opt_callback);
+    const options = this.getOptions_(optionsOrCallback);
+    const callback = /** @type {function(Event):boolean} */
+        (this.getCallback_(optionsOrCallback, opt_callback));
     let theme = options && options['theme'];
-    if (theme !== 'light' && theme !== 'dark') {
-      theme = 'light';
+    if (theme !== Theme.LIGHT && theme !== Theme.DARK) {
```
Comment (dvoytenko) [inline src/runtime/button-api.js]: I think we should move this code to `getOptions_`

File: src/runtime/runtime.js [hunk 33]
```diff
@@ -420,6 +420,13 @@ export class Runtime {
     return this.buttonApi_.create(optionsOrCallback, opt_callback);
   }
 
+  /** @override */
+  attachSmartButton(container, opt_options, callback) {
```
Comment (dvoytenko) [inline src/runtime/runtime.js]: Use `optionsOrCallback, opt_callback` naming here as well.

File: src/runtime/runtime.js [hunk 34]
```diff
@@ -816,6 +823,20 @@ export class ConfiguredRuntime {
     this.buttonApi_.attach(button, optionsOrCallback, opt_callback);
   }
 
+  /** @override */
+  attachSmartButton(container, optionsOrCallback, opt_callback) {
```
Comment (dvoytenko) [inline src/runtime/runtime.js]: s/container/button

File: src/runtime/runtime.js [hunk 35]
```diff
@@ -816,6 +823,20 @@ export class ConfiguredRuntime {
...
+    if (!isExperimentOn(this.win_, ExperimentFlags.SMARTBOX)) {
+      throw new Error('Not yet launched!');
+    }
+    if (!container) {
```
Comment (dvoytenko) [inline src/runtime/runtime.js]: not needed.

File: src/runtime/runtime.js [hunk 36]
```diff
@@ -816,6 +823,20 @@ export class ConfiguredRuntime {
...
+      throw new Error('No container element found to render Smart button!');
+    }
+    this.documentParsed_.then(() => {
```
Comment (dvoytenko) [inline src/runtime/runtime.js]: I'm actually not sure why we need to wait for `documentParsed` here. Can you explain?

File: src/runtime/smart-button-api.js [hunk 37]
```diff
@@ -0,0 +1,123 @@
...
+};
+
+
+/**
+ * The class for Smart button Api.
+ */
+export class SmartSubscriptionButtonApi {
+  /**
+   * @param {!./deps.DepsDef} deps
+   * @param {!Element} container
```
Comment (dvoytenko) [inline src/runtime/smart-button-api.js]: s/container/button

File: src/runtime/smart-button-api.js [hunk 38]
```diff
@@ -0,0 +1,123 @@
...
+      'publicationId': this.deps_.pageConfig().getPublicationId(),
+      'theme': options && options.theme || 'light',
+      'lang': this.options_ && this.options_.lang || 'en',
+    });
+  }
+
+  /**
+   * @return {!HTMLIFrameElement}
+   */
+  getElement() {
```
Comment (dvoytenko) [inline src/runtime/smart-button-api.js]: Do we need this method? If we do, imo it should return `button`, not iframe.

File: src/runtime/smart-button-api.js [hunk 39]
```diff
@@ -0,0 +1,123 @@
...
+        createElement(this.doc_, 'iframe', iframeAttributes));
+
+    /** @private @const {!Element} */
+    this.container_ = container;
+
+    /** @private {!../api/subscriptions.ButtonOptions|undefined} */
+    this.options_ = options;
+
+    /** @private const {?function()=} */
+    this.callback_ = callback;
```
Comment (dvoytenko) [inline src/runtime/smart-button-api.js]: Should we add a callback on the button itself right away as well to handle the keyboard tab/enter case?

File: assets/swg-button.css (hunk 24 above)
Comment (dparikh) [reply assets/swg-button.css]: Yes, Added.

File: assets/swg-button.css (hunk 23 above)
Comment (dparikh) [reply assets/swg-button.css]: Done

File: assets/swg-button.css (hunk 22 above)
Comment (dparikh) [reply assets/swg-button.css]: Done

File: assets/swg-button.css (hunk 25 above)
Comment (dparikh) [reply assets/swg-button.css]: Moved up.

File: examples/sample-pub/publisher.js (hunk 26 above)
Comment (dparikh) [reply examples/sample-pub/publisher.js]: Done

File: src/runtime/runtime.js (hunk 35 above)
Comment (dparikh) [reply src/runtime/runtime.js]: Was just extra check. Removed.

File: src/runtime/runtime.js (hunk 36 above)
Comment (dparikh) [reply src/runtime/runtime.js]: That was earlier implementation, when this was defined as a `flow`. Removed.

File: src/runtime/smart-button-api.js (hunk 37 above)
Comment (dparikh) [reply src/runtime/smart-button-api.js]: Done

File: src/runtime/smart-button-api.js (hunk 38 above)
Comment (dparikh) [reply src/runtime/smart-button-api.js]: Not really required, since used only within the class. Removed.

File: src/runtime/smart-button-api.js (hunk 39 above)
Comment (dparikh) [reply src/runtime/smart-button-api.js]: Added. Though, we need a way to `focus` the `iframe->button` when user `tabs` to the outer `button`.

File: src/api/subscriptions.js (hunk 27 above)
Comment (dparikh) [reply src/api/subscriptions.js]: Done

File: src/runtime/runtime.js (hunk 33 above)
Comment (dparikh) [reply src/runtime/runtime.js]: Done.

File: src/runtime/button-api.js (hunk 29 above)
Comment (dparikh) [reply src/runtime/button-api.js]: This is due to that fact that the returned `callback` function is passed as a param to `addEventListener` and if not typcasted, throws following error:

```
src/runtime/button-api.js:120: ERROR - Invalid type for parameter 2 of function button.addEventListener.
Expected : EventListener|function(Event):boolean|undefined|null
Found    : function(Event|null):boolean|null

    button.addEventListener('click', callback);
                                     ^^^^^^^^
1 error(s), 0 warning(s)```

Same function is used for `attachSmartButton` but not as an argument to `addEventListener` so needed this typecase.

File: src/runtime/button-api.js (hunk 30 above)
Comment (dparikh) [reply src/runtime/button-api.js]: This is to support two different function `attach` (having `addEventListener`) and 'attachSmartButton` (callback).
Without this, linter fails.

File: src/runtime/button-api.js (hunk 32 above)
Comment (dparikh) [reply src/runtime/button-api.js]: Re-factored this code to `getOptions_`

File: src/runtime/runtime.js (hunk 34 above)
Comment (dparikh) [reply src/runtime/runtime.js]: Done

File: src/runtime/button-api.js (hunk 31 above)
Comment (dparikh) [reply src/runtime/button-api.js]: Re-factored this, moved to `getOptions_`

File: src/runtime/button-api.js (hunk 28 above)
Comment (dparikh) [reply src/runtime/button-api.js]: Yes, updated.

Comment (dparikh) [reply src/runtime/button-api.js]: Yes. Remove that code updated for `options[lang]` as well.

File: examples/sample-pub/publisher.js [hunk 40]
```diff
@@ -104,6 +104,23 @@ function startFlowAuto() {
     });
     return;
   }
+
+  if (flow == 'smartbutton') {
+    whenReady(function(subsciptions) {
+      var subs = subsciptions;
+      whenDemoReady(function() {
+        var smartButton = document.querySelector('button.swg-smart-button');
```
Comment (dparikh) [inline examples/sample-pub/publisher.js]: Added stricts CSS selector to ensure that publisher is using `button` component.

File: src/runtime/smart-button-api.js [hunk 41]
```diff
@@ -0,0 +1,122 @@
...
+  /**
+   * Make a call to build button content and listens for the 'click' message.
+   * @return {!Element}
+   */
+  start() {
+    /**
+     * Add a callback to the button itself to fire the iframe click action when
+     * user tabs to the button and hits enter.
+     */
+    this.button_.addEventListener('click', this.callback_);
```
Comment (dvoytenko) [inline src/runtime/smart-button-api.js]: To come back to this point. It'd be better to indirect/wrap this callback than forcing the types all the way up. E.g.

```
const callback = this.callback;
button.addEventListener('click', () => {
  callback();
});
```

File: src/runtime/smart-button-api.js [hunk 42]
```diff
@@ -0,0 +1,122 @@
...
+    /** @private @const {!web-activities/activity-ports.ActivityPorts} */
+    this.activityPorts_ = deps.activities();
+
+    /** @private @const {!HTMLIFrameElement} */
+    this.iframe_ =
+    /** @type {!HTMLIFrameElement} */ (
+        createElement(this.doc_, 'iframe', iframeAttributes));
+
+    /** @private @const {!Element} */
+    this.button_ = button;
```
[the remark this answers is not in the document]
Comment (dparikh) [reply src/runtime/smart-button-api.js]: `assets/swg-button.css`

File: src/api/subscriptions.js [hunk 43]
```diff
@@ -236,6 +236,16 @@ export class Subscriptions {
    */
   attachButton(button, optionsOrCallback, opt_callback) {}
 
+  /**
+   * Attaches smartButton element and the provided callback.
+   * The default theme is "light".
+   *
+   * @param {!Element} container
```
[the remark this answers is not in the document]
Comment (dparikh) [reply src/api/subscriptions.js]: Done.

File: src/runtime/button-api.js [hunk 44]
```diff
@@ -98,28 +100,74 @@ export class ButtonApi {
...
+   * @param {!../api/subscriptions.ButtonOptions|function()} optionsOrCallback
+   * @return {!../api/subscriptions.ButtonOptions}
+   * @private
+   */
+  getOptions_(optionsOrCallback) {
+    const options = /** @type {!../api/subscriptions.ButtonOptions} */
+        (typeof optionsOrCallback != 'function' ?
+        optionsOrCallback : {'theme': Theme.LIGHT});
+
+    const theme = options && options['theme'];
```
[the remark this answers is not in the document]
Comment (dparikh) [reply src/runtime/button-api.js]: Yes. Removed additional check.

File: src/runtime/button-api.js (hunk 31 above)
Comment (dparikh) [reply src/runtime/button-api.js]: removed now.

File: src/runtime/smart-button-api.js [hunk 45]
```diff
@@ -0,0 +1,122 @@
...
+
+    /** @private {!../api/subscriptions.ButtonOptions|undefined} */
```
[the remark this answers is not in the document]
Comment (dparikh) [reply src/runtime/smart-button-api.js]: Removed.

File: src/runtime/smart-button-api.js (hunk 41 above)
Comment (dparikh) [reply src/runtime/smart-button-api.js]: Done

Comment (dparikh) [reply src/runtime/smart-button-api.js]: done

File: examples/sample-pub/views/article.html (hunk 21 above)
Comment (dparikh) [reply examples/sample-pub/views/article.html]: Done

File: src/runtime/button-api.js (hunk 29 above)
Comment (dparikh) [reply src/runtime/button-api.js]: Done.

File: src/runtime/button-api.js [hunk 46]
```diff
@@ -98,28 +100,75 @@ export class ButtonApi {
...
+   * @param {function()=} opt_callback
+   * @return {!Element}
+   */
+  attachSmartButton(deps, button, optionsOrCallback, opt_callback) {
+    const options = this.getOptions_(optionsOrCallback);
+    const callback = /** @type {function()} */
+        (this.getCallback_(optionsOrCallback, opt_callback));
+
+    // Add required CSS class, if missing.
+    if (!button.classList.contains('swg-smart-button')) {
```
Comment (dvoytenko) [inline src/runtime/button-api.js]: No need for `if`. It's implemented internally by `classList.add(...)`

File: src/runtime/smart-button-api.js [hunk 47]
```diff
@@ -0,0 +1,124 @@
...
+    this.callback_ = callback;
+
+    /** @private @const {string} */
+    this.src_ = feUrl('/smartboxiframe');
+
+    /** @private @const {!Object} */
+    this.args_ = feArgs({
+      'productId': this.deps_.pageConfig().getProductId(),
+      'publicationId': this.deps_.pageConfig().getPublicationId(),
+      'theme': options && options.theme || 'light',
```
Comment (dvoytenko) [inline src/runtime/smart-button-api.js]: Some `options` are resolved via arg and some via `this.options_`. Please make this consistent.

File: src/runtime/button-api.js (hunk 46 above)
Comment (dparikh) [reply src/runtime/button-api.js]: Removed.

File: src/runtime/smart-button-api.js (hunk 47 above)
Comment (dparikh) [reply src/runtime/smart-button-api.js]: Done.
